tools/inc/stljobs.h

```diff
@@ -0,0 +1,618 @@
+        validTill = 0;
+        start();
+        return result;
+    }
+
+    [[nodiscard]] HANDLE get_write_pipe() noexcept {
+        return writeHandle.get();
+    }
+
+private:
+    static void __stdcall callback(PTP_CALLBACK_INSTANCE, void* const thisRaw, void*, const ULONG ioResult,
+        const ULONG_PTR bytes, PTP_IO) noexcept {
+        switch (ioResult) {
+        case ERROR_SUCCESS:
+        case ERROR_OPERATION_ABORTED:
+            break;
+        default:
+            api_failure("StartThreadpoolIo + ReadFile callback", ioResult); // slams into noexcept
+            break;
+        }
+
+        const auto this_ = static_cast<output_collecting_pipe*>(thisRaw);
+        this_->validTill += bytes;
+        if (this_->running.load()) {
+            this_->read_some();
+        }
+    }
+
+    void ensure_target_buffer_space() {
+        if (bufferSize <= targetBuffer.size() - validTill) {
+            // already has enough space
+            return;
+        }
+
+        targetBuffer.resize(validTill + bufferSize);
+        targetBuffer.resize(std::min(static_cast<size_t>(ULONG_MAX), targetBuffer.capacity()));
+    }
+
+    void read_some() {
+        assert(running.load());
+        DWORD bytesRead = 0;
+        readIo.start_threadpool_io();
+        for (;;) {
+            ensure_target_buffer_space();
+            if (!ReadFile(readIo.get_file(), targetBuffer.data() + validTill,
+                    static_cast<unsigned long>(targetBuffer.size() - validTill), &bytesRead, &overlapped)) {
+                break;
+            }
+
+            validTill += bytesRead;
+        }
+
+        const auto lastError = GetLastError();
+        if (lastError == ERROR_IO_PENDING) {
+            return;
+        }
+
+        readIo.cancel_threadpool_io();
+        api_failure("ReadFile", lastError);
+    }
+
+    std::atomic<bool> running{};
+    std::string targetBuffer; // if running, owned by a threadpool thread, otherwise owned by the calling thread
+    size_t validTill{};
+    handle<invalid_handle_value_policy> writeHandle;
+    tp_io readIo;
+    OVERLAPPED overlapped{};
+};
+
+struct execution_result {
+    unsigned long exitCode;
+    std::string output;
+};
+
+class environment_block {
+public:
+    environment_block() {
+        default_block defaultBlock;
+        auto blockEnd = static_cast<const wchar_t*>(defaultBlock.blockRaw);
+        for (;;) {
+            const auto thisLen = wcslen(blockEnd);
+            blockEnd += thisLen;
+            ++blockEnd;
+
+            if (thisLen == 0) {
+                break;
+            }
+        }
+
+        block.assign(static_cast<const wchar_t*>(defaultBlock.blockRaw), blockEnd);
+    }
+
+    [[nodiscard]] void* get() noexcept {
+        return block.data();
+    }
+
+    void append_environment(const std::wstring_view key, const std::wstring_view value) {
+        block.reserve(block.size() + key.size() + value.size() + 2);

```

Comment:
This is quadratic time if called in a loop.

Reply:
For vector, it would be, but this is string. Always geometric with string :)

https://github.com/microsoft/STL/blob/cd8fee03209d3312bf3c79abf81371290f116d84/stl/inc/xstring#L3619